### ISSUE TITLE
Return the latest session_state on token refresh

### ISSUE THREAD

**Issue by cghislai** [keycloak/keycloak#31651]
### Description

Using keycloak 23 with a single user having multiple sessions in multiple tabs, then the session change iframe (at protocol/openid-connect/login-status-iframe.html) cannot be used, because

- the script in the iframe page checks the session state received from the parent page against the cookie, triggering a 'change' message to be sent to the parent page
- If the parent page refresh its token on 'change' event, the same session state is returned; even though the access token might be different

Since the session cookie is shared for all sessions across multiple tabs, and that each tab's RP may require storing the access token in session storage, or memory (thus not shared between tabs), then the same session state should be returned for all sessions of the same user on the same origin. 
Explained differently, the last session state set in the Set-Cookie response sent by keycloak should be returned for all following token refresh responses, provided the token was obtained for the same user and same origin.

In the openid session management specification (https://openid.net/specs/openid-connect-session-1_0.html), it says:

> In the case of an authorized Client (successful Authentication Response), the OP SHOULD change the value of the session state returned to the Client under one of the following events:
> 
> The set of users authenticated to the User Agent changes (login, logout, session add).
> The authentication status of Clients being used by the End-User changes.

While the library i use as the RP (angular-oauth2-oidc) does not follow the specification rigorously, having keycloak return the same session state on token refresh implies that the tokens must be shared between tabs, and that a locking mechanism must be implemented to avoid changing the session state in multiple tabs simultaneously.

### Motivation

- the session state is set in a keycloak cookie when the user logs in or off from keycloak. It represents the latest session_state returned by keycloak during a token response.
- the login-status-iframe.html checks the session state between that cookie and the value received in the message, which is session-dependant (tab-dependant)
- This cause inconsistencies for a single user on the same origin opening multiple session to keycloak.

### Details

When refreshing a token, instead of returning that specific session's 'session_state', i would return the latest one - the one that was last set in the Set-Cookie response header.

**Comment by jonkoops** [comment]
Keycloak 23 is no longer a supported version, can you verify this behaviour is still incorrect? I am not sure this is entirely correct in the first place. Keycloak doesn't allow a user to have multiple sessions AFAIK, only a single session per realm, not on a per-client basis.

**Comment by cghislai** [comment]
@jonkoops Ill test against latest version and confirm.

But keycloak definitely allows multiple sessions, even for the same client, at least in v23. In the screenshot below, multiple active sessions for the same user. The session id appears to be used to set the cookies and the 'session_state' field in the token responses.

![image](https://github.com/user-attachments/assets/349cdf03-ebc4-49e9-935a-761ecd052acb)

**Comment by jonkoops** [comment]
Hmmm, I have to admit that I am not entirely familiar with how multiple sessions are managed in Keycloak, even though I was under the impression that only a single session can exist per browser context. I'll ask around a bit.

**Comment by jonkoops** [comment]
So after asking around there should only be a single user session per context (e.g. a browser session), which would be a single cookie per session. The user can of course sign into multiple user sessions if they were to open a new browser, or a private window for example.

These user sessions can have multiple client sessions, which is when the user authenticates on a specific client. For example the user could sign into the 'foo' client and they will get a user session with a client session for 'foo', then when they sign into another client 'bar' that will create another client session that will be added to their existing user session, which shares a cookie.

The screenshot here shows a list of user sessions with their respective client sessions under the 'Clients' column. These are all separate user sessions, meaning they are logged in on different machines, or in different browser contexts on the same machine.

**Comment by cghislai** [comment]
Thanks for the explanation. The behaviour appears to be consistent with what you describes.

The issue I had was more subtle: 
- I open a new tab A, initiate a code flow, and get redirected to my app which exchange the code for TOKEN1, SESSION_ID_1
- I open another tab B, initiate a code flow, and get redirected to my app which exchange the code for TOKEN2, SESSION_ID_1

At this point, I only see a single session in keycloak indeed.

- I logout from tab B (using GET openid-connect/logout?id_token_hint=..), and get redirected to my app.

At this point, I don't see any session in keycloak, and my cookie is cleared.
In tab A, the absence of cookie causes the login-status-iframe.html to emit change; but the (active) token can still be exchanged for a new one, keeping the same session id SESSION_ID_1.
In the OP, I was re-logging again in tabB, and debugged tab A when the cookie was there with the updated session id.

I think I just need to implement logout backchannel correctly in my app, but I suspect something could be done on keycloak side as it continues to emit new token for SESSION_1 without setting the cookie nor listing the session in the UI.

Ill recheck this again against latest version and openid spec this evening/weekend.

**Comment by jonkoops** [comment]
> In tab A, the absence of cookie causes the login-status-iframe.html to emit change; but the (active) token can still be exchanged (refreshed) for a new one, keeping the same session id SESSION_ID_1.

Are you referring to the access token, refresh token, or both here? It indeed makes sense you could still use your access token on your service if it is not invalidated through back-channel logout, as the service has no way of knowing the token is no longer 'valid'. However, I would not expect a refresh token to be usable to renew a session on Keycloak if it has been signed out.

**Comment by cghislai** [comment]
> Are you referring to the access token, refresh token, or both here?

Refresh token. I use the token endpoint with grant_type: refresh_token; and get in response new tokens and the same session_state (which has been logged out in the other tab).

**Comment by mposolda** [comment]
@cghislai This is quite strange. After logout in tab2, the SSO session should be logged-out, which means the refresh token from tab1 should not be possible to use to refresh tokens.

It can be definitely good to update the latest version as there were some fixes in the meantime, which can be related (especially https://github.com/keycloak/keycloak/commit/f6071f68 ).

If you still see the issue, it can be helpful if you enable TRACE logging for the Keycloak events and you send snippet from the server log during the workflow you described above. Trace logging for the Keycloak events can be enabled for example when running server with something like this:

```
./kc.sh start --log-level="INFO,org.keycloak.events:TRACE"
```

**Comment by cghislai** [comment]
@mposolda @jonkoops After looking into this I don't think there is any issue in keycloak.

 I created a reproducer for my use-case at https://github.com/cghislai/keycloak-refresh-reproducer.

I noticed that keycloak creates 2 sessions when login in: one REGULAR, and one OFFLINE. 
My client was requesting the 'offline_access' scope, as per the library documentation. Relevant issue: https://github.com/manfredsteyer/angular-oauth2-oidc/issues/1241

When login out, only the REGULAR session was removed, but the OFFLINE session was kept active. Both sessions appear to share the same id. The OFFLINE sessions are displayed in the realm's session page, not the user's ones.

Revoking the OFFLINE session from keycloak  prevents obtaining new tokens using refresh token.
When the 'offline_access' scope is not requested/granted, no OFFLINE session is created.

Im closing this as it works as expected. Thank you for your input and sorry for the false alarm.
